src/cdomains/congruenceClosure.ml

```diff
@@ -0,0 +1,1275 @@
+        | Some (new_rep_s,z2) -> (* the successor already has a new representative, therefore we can just add it to the lookup map*)
+          uf, new_reps, closure new_cc [(successor_term, new_rep_s, Z.(old_z_s-z2))], reachable_old_reps
+        | None -> (* the successor state was not visited yet, therefore we need to find the new representative of the state.
+                     -> we choose a successor term *(t+z) for any t
+                     -> we need add the successor state to the list of states that still need to be visited
+                  *)
+          uf, TMap.add old_rep_s (successor_term, old_z_s) new_reps, new_cc, (old_rep_s, successor_term, old_z_s)::reachable_old_reps
+      else
+        (uf, new_reps, new_cc, reachable_old_reps)
+    | None ->
+      (* the term cannot be dereferenced, so we won't add this transition. *)
+      (uf, new_reps, new_cc, reachable_old_reps)
+  in
+  (* find all successors that are still reachable *)
+  let rec add_transitions uf new_reps new_cc = function
+    | [] -> new_reps, new_cc
+    | (old_rep, new_rep, z)::rest ->
+      let successors = LMap.successors old_rep cc.map in
+      let uf, new_reps, new_cc, reachable_old_reps =
+        List.fold (add_transition (old_rep, new_rep,z)) (uf, new_reps, new_cc, []) successors in
+      add_transitions uf new_reps new_cc (rest@reachable_old_reps)
+  in add_transitions uf new_reps new_cc
+    (List.unique_cmp ~cmp:(Tuple3.compare ~cmp1:(T.compare) ~cmp2:(T.compare) ~cmp3:(Z.compare)) reachable_old_reps)
+
+(** Find the representative term of the equivalence classes of an element that has already been deleted from the data structure.
+    Returns None if there are no elements in the same equivalence class as t before it was deleted.*)
+let find_new_root new_reps uf v =
+  match TMap.find_opt v new_reps with
+  | None -> uf, None
+  | Some (new_t, z1) ->
+    let t_rep, z2, uf = TUF.find uf new_t in
+    uf, Some (t_rep, Z.(z2-z1))
+
+let remove_terms_from_diseq diseq new_reps cc =
+  let disequalities = Disequalities.get_disequalities diseq
+  in
+  let add_disequality (uf,new_diseq) = function
+    | Nequal(t1,t2,z) ->
+      begin match find_new_root new_reps uf t1 with
+        | uf, Some (t1',z1') ->
+          begin match find_new_root new_reps uf t2 with
+            | uf, Some (t2', z2') -> uf, (t1', t2', Z.(z2'+z-z1'))::new_diseq
+            | _ -> uf, new_diseq
+          end
+        | _ -> uf, new_diseq
+      end
+    | _-> uf, new_diseq
+  in
+  let uf,new_diseq = List.fold add_disequality (cc.uf,[]) disequalities
+  in congruence_neq {cc with uf} new_diseq
+
+let remove_terms_from_bldis bldis new_reps cc =
+  let uf_ref = ref cc.uf in
+  let find_new_root_term t =
+    let uf, new_root = find_new_root new_reps !uf_ref t in
+    uf_ref := uf;
+    Option.map fst new_root in
+  let bldis = BlDis.filter_map_lhs find_new_root_term bldis in
+  !uf_ref, BlDis.filter_map find_new_root_term bldis
```

Comment:
What is the intention here? That `uf_ref` is first returned, and then the `filter_map find_new_root_term` modifies `uf_ref` on the side, but without effect?

Reply:
I adapted the code so that it is clearer the that `BlDis.filter_map` function in the last line is executed first.